### ISSUE TITLE
foreign key properties

### ISSUE THREAD

**Issue by dimitri-yatsenko** [datajoint/datajoint-python#301]
I propose to allow specification of foreign key properties `unique` and `required`.

I propose the following syntax.

```
-> [unique, nullable] lab.Animal 
```

"Unique" means that each animal in `lab.Animal` can be referenced only once from the referencing table.

"Optional" means that the foreign key values can be null.  When null, the unique constraint is not applicable. 

By default, foreign keys are not unique and not optional. 

This syntax is compatible with the syntax for renamed foreign keys:

```
(subject_id) -> [unique] lab.Animal
```

**Comment by Visdoom** [comment]
This is exactly what I need. Please put it back into the road map. Is there any progress on this?

**Comment by dimitri-yatsenko** [comment]
This feature is described in http://docs.datajoint.io/data-definition/Foreign-keys.html

**Comment by Visdoom** [comment]
Hey Dimitri, 
I got the new release and added foreign key renaming to my tables.
However, now I am trying to populate the calculated table with new values ( there are some inserted already ) and the populate method returns without any execution of jobs.

The 'error' occurs in the `populate` method in line 92:
`todo -= self.target`

Even though todo contains ~600 entries before and target contains maybe 10 the result of this line is an empty `todo` object.
Could it be, that the mismatch in keys of `todo` and `target` cause that misbehaviour? 
`Todo` works with the original heading of the foreign key ( in my case _id_ ) whereas `target` operates on the renamed version of it.

**Comment by eywalker** [comment]
Hi @Visdoom, could you tell us how many entries are found if you run 1) `todo - self.target` and 2) `todo & self.target`? Also, it would be helpful to know what are the primary keys for `todo` and `target` exactly. I'm suspecting that `todo` and `target` must be sharing perhaps all but one primary key, and because `id` was renamed, `self.target` no longer restricts the `todo` appropriately.

**Comment by dimitri-yatsenko** [comment]
Yes, we need to update the default `key_source` to work correctly for
computed tables with renamed foreign keys.  Let's write this up as an
issue.  At this point, you will need to overwrite the key source with the
right query.  If you need help, please paste the definition of the your
computed table.

**Comment by Visdoom** [comment]
Hey @eywalker, @dimitri-yatsenko,
I changed back to not renaming the key. But you are right, overwriting the key_source query would probably have helped too.

As to the questions:
`todo - self.target` resulted in 0 entries.
I guess `todo & self.target` would have been the sum (~610). 
Specifically, the renaming was
`(fmparam_id) -> FMParameterSet(id)`

**Comment by dimitri-yatsenko** [comment]
Renamed foreign keys are rarely strictly necessary.  Unfortunately, it is a common practice in non-DataJoint databases to always rename.  This makes queries more difficult.  We need to update the cascaded delete (#300) and the default key source to process them correctly.

**Comment by dimitri-yatsenko** [comment]
By the way, this specific issue deals with foreign key options `unique` and `nullable`, not renamed keys specifically.  We can write this as a new issue or add to #300.

**Comment by eywalker** [comment]
Yes - I have already created an issue to track the renamed keys and `key_source` at #345

**Comment by dimitri-yatsenko** [comment]
This was closed prematurely, but it's okay, we are about to complete the implementation.
